File: examples/dijkstra/Go/dijkstra.go
// dijkstra.go - Dijkstra's shortest path algorithm implementation

package main

import (
	"math"

	dheap "github.com/PCfVW/d-Heap-priority-queue/Go/v2/src"
)

// Infinity represents an unreachable distance.
const Infinity = math.MaxInt

// Dijkstra finds the shortest paths from a source vertex to all other vertices
// using a d-ary heap priority queue.
//
// Parameters:
//   - graph: The input graph with vertices and weighted edges
//   - source: The source vertex to find shortest paths from
//   - d: The arity of the heap (default 4 for optimal performance)
//
// Returns a DijkstraResult containing distances and predecessors for path reconstruction.
func Dijkstra(graph Graph, source string, d int) DijkstraResult {
	return dijkstraImpl(graph, source, d, nil)
}

// DijkstraInstrumented runs the algorithm with comparison-count instrumentation
// attached to the underlying priority queue. Returns the result alongside a
// pointer to the populated dheap.Stats so the caller can read per-operation
// comparison counts (Insert / Pop / IncreasePriority / DecreasePriority /
// UpdatePriority — only Insert / Pop / IncreasePriority are exercised by this
// algorithm).
func DijkstraInstrumented(graph Graph, source string, d int) (DijkstraResult, *dheap.Stats) {
	stats := &dheap.Stats{}
	result := dijkstraImpl(graph, source, d, stats)
	return result, stats
}

// dijkstraImpl is the shared algorithm body. When stats is nil, the heap runs
// with no instrumentation; when non-nil, every comparison is bucketed into the
// matching operation counter inside *stats.
func dijkstraImpl(graph Graph, source string, d int, stats *dheap.Stats) DijkstraResult {
	// Build adjacency list for efficient neighbor lookup
	adjacency := make(map[string][]struct {
		To     string
		Weight int
	})
	for _, v := range graph.Vertices {
		adjacency[v] = nil
	}
	for _, e := range graph.Edges {
		adjacency[e.From] = append(adjacency[e.From], struct {
			To     string
			Weight int
		}{To: e.To, Weight: e.Weight})
	}

	// Initialize distances and predecessors
	distances := make(map[string]int)
	predecessors := make(map[string]*string)

	// Create priority queue with min-heap by distance
	pq := dheap.New(dheap.Options[Vertex, string]{
		D:            d,
		Comparator:   dheap.MinBy(func(v Vertex) int { return v.Distance }),
		KeyExtractor: func(v Vertex) string { return v.ID },
		Stats:        stats, // nil == no instrumentation
	})

	// Set initial distances and add to priority queue
	for _, v := range graph.Vertices {
		distance := Infinity
		if v == source {
			distance = 0
		}
		distances[v] = distance
		predecessors[v] = nil
		pq.Insert(Vertex{ID: v, Distance: distance})
	}

	// Main algorithm loop
	for !pq.IsEmpty() {
		current, _ := pq.Pop()

		// Skip if we've already found a shorter path
		if current.Distance > distances[current.ID] {
			continue
		}

		// Skip if current distance is infinity (unreachable)
		if current.Distance == Infinity {
			continue
		}

		// Check all neighbors
		for _, neighbor := range adjacency[current.ID] {
			newDistance := current.Distance + neighbor.Weight

			if newDistance < distances[neighbor.To] {
				distances[neighbor.To] = newDistance
				pred := current.ID
				predecessors[neighbor.To] = &pred

				// Update priority in queue
				// In a min-heap, decreasing distance = increasing priority (more important)
				if pq.Contains(Vertex{ID: neighbor.To}) {
					pq.IncreasePriority(Vertex{ID: neighbor.To, Distance: newDistance})
				}
			}
		}
	}

	return DijkstraResult{
		Distances:    distances,
		Predecessors: predecessors,
	}
}

// ReconstructPath builds the shortest path from source to target using predecessors.
//
// Returns the path as a slice of vertex IDs, or nil if no path exists.
func ReconstructPath(predecessors map[string]*string, source, target string) []string {
	if predecessors[target] == nil && target != source {
		return nil // No path exists
	}

	var path []string
	current := &target

	for current != nil {
		path = append([]string{*current}, path...) // Prepend
		current = predecessors[*current]
	}

	if len(path) > 0 && path[0] == source {
		return path
	}
	return nil
}

File: examples/dijkstra/Rust/src/dijkstra.rs
//! dijkstra.rs - Dijkstra's shortest path algorithm implementation

use crate::types::{DijkstraResult, Graph, Vertex};
use d_ary_heap::{MinBy, PriorityQueue};
use std::collections::HashMap;

/// Infinity represents an unreachable distance.
pub const INFINITY: i32 = i32::MAX;

/// Dijkstra's shortest path algorithm using a d-ary heap priority queue.
///
/// Finds the shortest paths from a source vertex to all other vertices in a weighted
/// graph with non-negative edge weights.
///
/// # Arguments
///
/// * `graph` - The input graph with vertices and weighted edges
/// * `source` - The source vertex to find shortest paths from
/// * `d` - The arity of the heap (typically 4 for optimal performance)
///
/// # Returns
///
/// A `DijkstraResult` containing distances and predecessors for path reconstruction.
pub fn dijkstra(graph: &Graph, source: &str, d: usize) -> DijkstraResult {
    // Build adjacency list for efficient neighbor lookup
    let mut adjacency: HashMap<String, Vec<(String, i32)>> = HashMap::new();
    for vertex in &graph.vertices {
        adjacency.insert(vertex.clone(), Vec::new());
    }
    for edge in &graph.edges {
        adjacency
            .get_mut(&edge.from)
            .unwrap()
            .push((edge.to.clone(), edge.weight));
    }

    // Initialize distances and predecessors
    let mut distances: HashMap<String, i32> = HashMap::new();
    let mut predecessors: HashMap<String, Option<String>> = HashMap::new();

    // Create priority queue with min-heap by distance
    let mut pq = PriorityQueue::new(d, MinBy(|v: &Vertex| v.distance)).unwrap();

    // Set initial distances and add to priority queue
    for vertex in &graph.vertices {
        let distance = if vertex == source { 0 } else { INFINITY };
        distances.insert(vertex.clone(), distance);
        predecessors.insert(vertex.clone(), None);
        pq.insert(Vertex {
            id: vertex.clone(),
            distance,
        });
    }

    // Main algorithm loop
    while !pq.is_empty() {
        let current = pq.pop().unwrap();

        // Skip if we've already found a shorter path
        if current.distance > *distances.get(&current.id).unwrap() {
            continue;
        }

        // Skip if current distance is infinity (unreachable)
        if current.distance == INFINITY {
            continue;
        }

        // Check all neighbors
        if let Some(neighbors) = adjacency.get(&current.id) {
            for (neighbor_id, weight) in neighbors {
                let new_distance = current.distance + weight;

                if new_distance < *distances.get(neighbor_id).unwrap() {
                    distances.insert(neighbor_id.clone(), new_distance);
                    predecessors.insert(neighbor_id.clone(), Some(current.id.clone()));

                    // Update priority in queue
                    // In a min-heap, decreasing distance = increasing priority (more important)
                    let neighbor_vertex = Vertex {
                        id: neighbor_id.clone(),
                        distance: 0, // dummy value for contains check
                    };
                    if pq.contains(&neighbor_vertex) {
                        pq.increase_priority(&Vertex {
                            id: neighbor_id.clone(),
                            distance: new_distance,
                        })
                        .unwrap();
                    }
                }
            }
        }
    }

    DijkstraResult {
        distances,
        predecessors,
    }
}

/// Reconstructs the shortest path from source to target using predecessors.
///
/// Builds the path by following predecessor links backwards from the target,
/// then reversing the result.
///
/// # Arguments
///
/// * `predecessors` - Predecessor map from dijkstra result
/// * `source` - Source vertex
/// * `target` - Target vertex
///
/// # Returns
///
/// A vector of vertex IDs representing the path, or `None` if no path exists.
pub fn reconstruct_path(
    predecessors: &HashMap<String, Option<String>>,
    source: &str,
    target: &str,
) -> Option<Vec<String>> {
    if predecessors.get(target)?.is_none() && target != source {
        return None; // No path exists
    }

    let mut path = Vec::new();
    let mut current = Some(target.to_string());

    while let Some(vertex) = current {
        path.push(vertex.clone());
        current = predecessors.get(&vertex)?.clone();
    }

    path.reverse();

    if !path.is_empty() && path[0] == source {
        Some(path)
    } else {
        None
    }
}
